Objects/longobject.c

```diff
@@ -4366,23 +4371,65 @@ long_pow(PyObject *v, PyObject *w, PyObject *x)
         }
     }
     else {
-        /* Left-to-right 5-ary exponentiation (HAC Algorithm 14.82) */
-        Py_INCREF(z);           /* still holds 1L */
-        table[0] = z;
-        for (i = 1; i < 32; ++i)
-            MULT(table[i-1], a, table[i]);
+        /* Left-to-right k-ary sliding window exponentiation
+         * (HAC Algorithm 14.85)
+         */
+        Py_INCREF(a);
+        table[0] = a;
+        MULT(a, a, a2);
+        /* table[i] == a**(2*i + 1) % c */
+        for (i = 1; i < EXP_TABLE_LEN; ++i)
+            MULT(table[i-1], a2, table[i]);
+        Py_CLEAR(a2);
+
+        /* Repeatedly extract the next (no more than) EXP_WINDOW_SIZE bits
+         * into `pending`, starting with the next 1 bit.  The current bit
+         * length of `pending` is `blen`, and if `pending` is non-zero the
+         * current number of trailing zeroes `ntz`.
+         */
+        int pending = 0, blen = 0, ntz = 0;
+#define ABSORB_PENDING  do { \
```

Comment:
Consider making this a static inline function that passes in the addresses of the variables.

Reply:
Actively don't want to do that. This block is expanded exactly twice, will never be used anywhere else, and doesn't correspond to a coherent abstraction. It's just doing exactly what needs to be done in the two places it's expanded, and is a macro solely to minimize the chance of the two code blocks getting out of synch. Indeed, I'd rather type it out in full twice than try to pretend it's a coherent "function", full of noisy `*pending` and `*ntz`  and `*blen` obfuscations.

Reply:
BTW, note too that `ABSORB_PENDING` contains instances of the `MULT` macro, which in turn contains use of the `REDUCE` macro, and both of the latter contain `goto Error;` statements. Those cannot be turned into functions, or used by other functions, without significant surgery to rework them (and also correspondingly change _all_ the code sites that use them).

More positively, I actively like having the body of `ABSORB_PENDING` so physically close to the small loop nest that uses that macro. They're exceedingly tightly coupled in what they do, and it would be a major PITA to have to scroll the editor multiple pages (`long_pow()` is a long function!) to flip between them.